src/rollup/index.ts

```diff
@@ -211,45 +212,81 @@ export default function rollup(rawInputOptions: GenericConfigObject): Promise<Ro
-									return chunk.render(outputOptions, addons, outputChunk).then(rendered => {
-										outputChunk.code = rendered.code;
-										outputChunk.map = rendered.map;
-
-										return graph.pluginDriver.hookParallel('ongenerate', [
-											{ bundle: outputChunk, ...outputOptions },
-											outputChunk
-										]);
-									});
+									const facadeModule = chunk.facadeModule;
+									const chunkInfo = {
+										dynamicImports: chunk.getDynamicImportIds(),
+										exports: chunk.getExportNames(),
+										facadeModuleId: facadeModule && facadeModule.id,
+										imports: chunk.getImportIds(),
+										isDynamicEntry:
+											facadeModule !== null && facadeModule.dynamicallyImportedBy.length > 0,
+										isEntry: facadeModule !== null && facadeModule.isEntryPoint,
+										modules: chunk.renderedModules,
+										get name() {
+											return chunk.getChunkName();
+										}
+									} as PreRenderedChunk;
+									return graph.pluginDriver
+										.hookReduceValue(
+											'augmentChunkHash',
+											'',
+											[chunkInfo],
+											(hashForChunk, pluginHash) => {
+												if (pluginHash) {
+													hashForChunk += pluginHash;
+												}
+												return hashForChunk;
+											}
+										)
+										.then(hashForChunk => {
+											if (hashForChunk) {
+												chunk.augmentedHash = hashForChunk;
+											}
+										});
 								})
-							).then(() => {});
+							).then(() => {
+								assignChunkIds(chunks, inputOptions, outputOptions, inputBase, addons);
+
+								// assign to outputBundle
+								for (let i = 0; i < chunks.length; i++) {
+									const chunk = chunks[i];
+									const facadeModule = chunk.facadeModule;
+
+									outputBundle[chunk.id] = {
+										code: undefined as any,
+										dynamicImports: chunk.getDynamicImportIds(),
+										exports: chunk.getExportNames(),
+										facadeModuleId: facadeModule && facadeModule.id,
+										fileName: chunk.id,
+										imports: chunk.getImportIds(),
+										isDynamicEntry:
+											facadeModule !== null && facadeModule.dynamicallyImportedBy.length > 0,
+										isEntry: facadeModule !== null && facadeModule.isEntryPoint,
+										map: undefined,
+										modules: chunk.renderedModules,
+										get name() {
+											return chunk.getChunkName();
+										}
+									} as OutputChunk;
+								}
+
+								return Promise.all(
+									chunks.map(chunk => {
+										const outputChunk = outputBundle[chunk.id] as OutputChunk;
+										return chunk.render(outputOptions, addons, outputChunk).then(rendered => {
+											outputChunk.code = rendered.code;
+											outputChunk.map = rendered.map;
+
+											return graph.pluginDriver.hookParallel('ongenerate', [
+												{ bundle: outputChunk, ...outputOptions },
+												outputChunk
+											]);
+										});
+									})
+								).then(() => {});
+							});
```

Comment:
I am not sure what exactly causes the test failures, e.g.  misc/'handles different import paths for different outputs', but it is definitely caused by some of the changes in this file. Maybe it becomes clearer once the code is deduplicated. My guess is that there is something that has a side effect that does not look side-effectful here.

Reply:
I have found what's causing the errors and its pretty strange:

```js
return graph.pluginDriver
  .hookParallel('renderStart', [])
  .then(() => createAddons(graph, outputOptions))
  .then(addons => {
    // ...
    return Promise.all(
      chunks.map(chunk => {
        const outputChunk = outputBundle[chunk.id] as OutputChunk;
        return chunk.render(outputOptions, addons, outputChunk).then(rendered => {
          // ...  
          return graph.pluginDriver.hookParallel('ongenerate', [
            { bundle: outputChunk, ...outputOptions },
           outputChunk
          ]);
        });
      })
    ).then(() => {});
})
```
Works fine and test pass whereas if I move the `Promise.all` block into another promise:

```js
return graph.pluginDriver
  .hookParallel('renderStart', [])
  .then(() => createAddons(graph, outputOptions))
  .then(addons => {
    // ...
    const augmentChunkHashes = Promise.resolve();
    return augmentChunkHash.then( () => {
      return Promise.all(
        chunks.map(chunk => {
          const outputChunk = outputBundle[chunk.id] as OutputChunk;
          return chunk.render(outputOptions, addons, outputChunk).then(rendered => {
            // ...  
            return graph.pluginDriver.hookParallel('ongenerate', [
              { bundle: outputChunk, ...outputOptions },
              outputChunk
            ]);
          });
        })
      ).then(() => {});
   })
})
```
Some tests break but if I understand correctly:

```js
Promise.all(promises)
  .then(()=>{}) 
=== 
Promise.resolve()
  .then(()=>Promise.all(promises).then(()=>{}))

Reply:
There is a slight difference in that errors thrown synchronously when generating the promises will be caught and reject the Promise in the second case while in the first case, they will be thrown synchronously. But I do not think this is the issue here.

I cannot seem to be able to confirm your finding, will do some digging myself.

Reply:
Also in the second case, execution will be one micro-tick slower, which would point to a possible race condition. But I really hope this is not the issue.

Reply:
Ok, I understand the issue now. It could be argued that this is a bug in Rollup but it is actually only a bug with your modification.
The problem is that the core part of chunk rendering is stateful, i.e. you cannot just interleave the renderings of two different chunks as they leave information about the current rendering process at variables etc. I know this is not ideal but for the time being, I fear we will need to live with this.

The state only needs to be preserved between `chunk.preRender` and `chunk.render`. Previously, this was guaranteed but with your modification, there is now an asynchronous micro-tick between those two calls where another output rendering can sneak in. Which is a problem as the CLI does indeed use `Promise.all` to render outputs.

Reply:
Honestly not sure what a good solution is except forbidding `augmentChunkHash` to be asynchronous for now.

Reply:
Thanks for looking into this, I was loosing my mind. So if I get it correctly the correct order of events would be to `preRender` and then `render` the chunk for each chunk but with the constraint that this sequence has to be completed before another output goes in. Moving the rendering to another micro-task makes it possible for some sequences when several outputs are generated to go `preRender-ChunkA-OutA` `preRender-ChunkA-OutB ` `render-ChunkA-OutA` or something among those lines breaking the sequence since they are generated inside a `Promise.all` in parallel.

Another possible solution could be running this hook before just before `generate` but I would rather have the information generated during `preRender` and make this `syncrhonous` with a new `hookReduceValueSync` hook for example.